examples/networkgraph.cpp

```diff
@@ -0,0 +1,909 @@
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <getopt.h>
+#include <sstream>
+
+#include <opm/io/eclipse/ERst.hpp>
+
+#include <opm/input/eclipse/Deck/Deck.hpp>
+
+#include <opm/input/eclipse/Parser/ErrorGuard.hpp>
+#include <opm/input/eclipse/Parser/InputErrorAction.hpp>
+#include <opm/input/eclipse/Parser/Parser.hpp>
+#include <opm/input/eclipse/Parser/ParseContext.hpp>
+
+
+class Node
+{
+public:
+    Node(const std::string& name);
+
+    std::string name() const { return m_name; }
+    void set_outlet(std::shared_ptr<Node> outlet) { m_outlet = outlet; };
+    std::shared_ptr<Node> get_outlet() { return m_outlet; };
+
+    void add_inlet_node(std::shared_ptr<Node> n);
+
+    void set_vfp(int vfp){m_vfp = vfp;};
+    int get_vfp(){ return m_vfp;};
+
+    void set_fixed_pres(double pres){m_fixed_pres = pres; };
+    double get_fixed_pres(){return m_fixed_pres; };
+
+    int get_xpos(){return m_xpos;};
+
+    void print(std::stringstream& netw_str);
+
+    bool detete_from_inlet_list(const std::string& name);
+
+    bool add_well(const std::string& name);
+
+    void reset_outlet(){m_outlet = nullptr;};
+
+private:
+    int m_vfp;
+    int m_xpos = -1;
+    double m_fixed_pres = -1.0;
+    int m_ind = -1;
+
+    std::string m_name;
+    std::shared_ptr<Node> m_outlet;
+    std::vector<std::shared_ptr<Node>> m_inlet_list;
+    std::vector<std::string> m_well_list;
+
+    std::shared_ptr<Node> next_branch();
+};
+
+
+Node::Node(const std::string& name)
+   : m_name(name)
+{
+    m_vfp = -1;
+}
+
+std::shared_ptr<Node> Node::next_branch(){
+
+    std::shared_ptr<Node> p1 = m_outlet;
+
+    while (p1 != nullptr){
+        for (size_t n = 0; n < p1->m_inlet_list.size(); n ++){
+            if (p1->m_inlet_list[n]->m_xpos == -1){
+                return p1;
+            }
+        }
+        p1 = p1->m_outlet;
+    }
+
+    return p1;
+}
+
+bool Node::detete_from_inlet_list(const std::string& name){
+
+    int ind = -1;
+
+    for (size_t n = 0; n < m_inlet_list.size(); n ++)
+        if (m_inlet_list[n]->name() == name)
+            ind = n;
+
+    if (ind < 0){
+        return false;
+    } else {
+        m_inlet_list.erase (m_inlet_list.begin() + ind);
+        return true;
+    }
+}
+
+bool Node::add_well(const std::string& name){
+   if (m_inlet_list.size() > 0)
+       return false;
+
+   m_well_list.push_back(name);
+
+   return true;
+}
+
+
+void Node::print(std::stringstream& netw_str){
+
+    netw_str.seekg(0, std::ios::end);
+    int init_length = netw_str.tellg();
+
+    int pos_lineshift = 0;
+
+    auto p = netw_str.str().find_last_of("\n");
```

Comment:
I think the str() method may be quite expensive, as it probably constructs a complete string every time. So invoking it multiple times could be a performance killer, although it might not matter for the typical use case for this app.

I am fine with leaving it in if you do not want to change it.

Reply:
The str() method is called twice for each node. First I need to find position of end of line, then I'm adding more characters to the string stream before I again need to find the new position.  I would like to keep this as is until we see that this is a performance issue.